### ISSUE TITLE
The optimizer needs some refactorization

### ISSUE THREAD

**Issue by oysteijo** [oysteijo/simd_neuralnet#67]
It's a little bit inconvenient and clumsy to use. As an example: You have to specify the epoch_runner function - but that should be given by the implementation. For the end user it should be like:
```c
optimizer_t *opt = OPTIMIZER( adamw_new( nn, ADAMW_SETTINGS( ) ) ); 
optimizer_t *opt = OPTIMIZER( SGD_new( nn, SGD_SETTINGS( ) ) ); 
```
So, each `<name>_new()` function should return a optimizer implementation cat can be casted to the abstract. These functions should set the epoch runners etc.

**Comment by oysteijo** [comment]
Look at how this is done in `callback.h`

**Comment by oysteijo** [comment]
So, an important question becomes:

Which data fields should be in the abstract optimizers, and which should be in the specific implemetation? Here is what I believe, however I might change my mind. Nothing is carved in stone.

* the neural network (nn) is part of all optimizers and must be passed whatever. A reference to the neural network must be in the abstract optimizer.
* `interations` (just a counting number) it can be in the abstract
* `shuffle` (boolean variable) This is about the dataset and whether to shuffle this or not. Can be in abstract - actually *must* be in abstract.
* `batchsize` (integer) It's the same for all optimizers. Can also be in abstract - actually *must* be in abstract.
* `progress` (function pointer) - From a technical point it doesn't matter, but I think it belong (logically) in abstract
* `metrics` (array of function pointers) - This is used in the template code, so it must be in the abstract.
* `pivot` (array of ints) - This is hidden from the end user - should be in abstract

But...
* `learning_rate` ? It feels naturally to have this in the abstract, but what then when get a implementation that does not use a learning rate?
* `decay`? Doesn't make sense to have it in the abstract unless the learning rate is in the abstract.
* `momentum`  float value ? - Does all optimizers have this?
* `nesterov` boolean ? - Does all optimizers support Nesterov momentum.
* Other $\beta_1$, $\beta_2$, $\rho$ and so on.

Still up for some discussion

**Comment by oysteijo** [comment]
Some comments to the above:

`iterations` is only used when decay for the learning rate is used. It can hence be moved as a hidden data field in the spesific optimizers what supports weight decay. I also suggest it should be called `n_iterations`

`shuffle`, `batchsize`, `progress` and `metrics` should be accessable with get/set ? Or at least be able to specify at creation.

`pivot` should be in abstract and internal. No one should be able to access it.

So, how do vi make configuration macro(s). Should we have one macro for the abstract and one for the implementation?

```c
optimizer_t *opt = OPTIMIZER( SGD_new( nn, OPTIMIZER_SETTINGS(), SGD_SETTINGS( ) ) ); 
``` 
Or should we try to connect them in a way?

**Comment by oysteijo** [comment]
Digression 

The parameters or properties of each instance of an optimizer should be named.... or rather the macro holding them. So what do we call them?

* Options
* Settings
* Configuration
* Properties
* Parameters
* Preferences 

So, I asked a chat bot....

The terms options, settings, configuration, properties, parameters, and preferences are often used in the context of software and technology, and while they are related, they do have different meanings and implications. Here's a brief overview of each term:

Options:
Options generally refer to individual choices or selections that can be made within a software application or system. They often represent specific features or functionalities that can be turned on or off, adjusted, or customized to suit the user's needs or preferences. Options are typically presented to the user in the form of checkboxes, radio buttons, dropdown lists, or other user interface elements.

Settings:
Settings are the configuration details that determine how a software application or system behaves. They encompass a broad range of configurable parameters, including options, preferences, and other customization choices. Settings are often used to control the overall behavior or appearance of the software, and they are typically organized into categories or sections within the user interface to make them more manageable.

Configuration:
Configuration refers to the arrangement and customization of the components and settings of a software application or system to make it suitable for a specific purpose or environment. This can include setting up connections to external services, specifying default values for certain options, defining the behavior of the software under different conditions, and more. Configuration is often performed during the initial setup of the software and may be adjusted as needed over time.

Properties:
In the context of software development, properties often refer to the characteristics or attributes of an object or component within a program. For example, a graphical user interface element might have properties such as its size, position, color, visibility, and so on. These properties can be manipulated through code or through a user interface to control the appearance and behavior of the software.

Parameters:
Parameters are values that are passed to a function, method, or program in order to customize its behavior or provide input for its operation. In software development, parameters are used to enable code reusability and flexibility by allowing the same function or method to be used with different inputs. Parameters can be thought of as variables that are used to configure or influence the execution of a particular piece of code.

Preferences:
Preferences are individual user choices that dictate how a software application should behave or appear. They often encompass a wide range of settings and options that are specific to each user, such as language preferences, display settings, notification preferences, and so on. Preferences are designed to allow users to personalize their experience with the software to better suit their individual needs and preferences.

While these terms are related and sometimes used interchangeably, they each have specific connotations and roles within the context of software and technology.

**Comment by oysteijo** [comment]
I'm not sure it made me wiser. However.... I am getting new possible terms:

* Attributes
* Fields
* Members

Hmm...

**Comment by oysteijo** [comment]
... or 'setup'?

Of all these I think I actually likes best **properties** and **parameters**. The drawback is the length, in that case I can settle for **config**.  Let me try the syntax here:

```c
optimizer_t *opt = OPTIMIZER( SGD_new( nn, OPTIMIZER_PROPERTIES(), SGD_PROPERTIES( ) ) ); 

optimizer_t *opt = OPTIMIZER( SGD_new( nn, OPTIMIZER_PARAMETERS(), SGD_PARAMETERS( ) ) ); 

optimizer_t *opt = OPTIMIZER( SGD_new( nn, OPTIMIZER_CONFIG(), SGD_CONFIG( ) ) ); 
```
The line length here is 94 characters with a 4 space indent for parameters and properties. Config is 4 letters less ant it's written twice, so the line length using config becomes 86 (including 4 space indent)

Let's also look at this with some parameters and line breaks:
```c
optimizer_t *opt = OPTIMIZER(
    SGD_new( nn, 
        OPTIMIZER_PROPERTIES( .shuffle=true, batchsize=32 ),
        SGD_PROPERTIES( .nesterov=true, momentum=0.95 ) )
    ); 

optimizer_t *opt = OPTIMIZER(
    SGD_new( nn, 
        OPTIMIZER_PARAMETERS( .shuffle=true, batchsize=32 ),
        SGD_PARAMETERS( .nesterov=true, momentum=0.95 ) )
    ); 

optimizer_t *opt = OPTIMIZER(
    SGD_new( nn, 
        OPTIMIZER_CONFIG( .shuffle=true, batchsize=32 ),
        SGD_CONFIG( .nesterov=true, momentum=0.95 ) )
    ); 

```

**Comment by oysteijo** [comment]
Based on the above, I think I go with **properties**.

Then the next - what to do with the callback. I guess we should be consistent. Let me do the optimizer first. Then think about the consistence.

**Comment by oysteijo** [comment]
See! It doesn't look that bad:
```c
    optimizer_t *opt = OPTIMIZER(
         SGD_new(
             nn, 
             OPTIMIZER_PROPERTIES(
                .batchsize = 16,
                .shuffle   = true,
                .metrics   = ((metric_func[]){ get_metric_func( get_loss_name( nn->loss ) ),
                    get_metric_func( "categorical_accuracy" ), NULL }),
                /* .progress  = NULL */
            ),
            SGD_PROPERTIES( .learning_rate=learning_rate )
         )
    );
```

Now I also have to consider `.metrics`. Does it make any sense to have none? Does it make sense to have a NULL pointer as the mertics? Or the first element to be a NULL pointer? In case it doesn't make sense - should we then sneak in metric corresponding to the neural network loss function? And in case "yes" to that, how do we sneak it in?

**Comment by oysteijo** [comment]
A `METRIC_LIST()` macro? And a `METRIC_FROM_NEURALNET()`?
```c
#define METRIC_FROM_NEURALNET(nnet) get_metric_func( get_loss_name( nnet->loss )
#define METRIC_LIST(...) ((metric_func[]){ __VA_ARGS__, NULL })

/* ... some code ... */

.metrics   = METRIC_LIST(METRIC_FROM_NEURALNET(nn), get_metric_func( "categorical_accuracy" )),

```
Something like that?

But then ..... it doesn't make any sense to have no metrics at all, does it? Say the OPTIMIZER_PROPERTIES macro sets the default metric to `NULL`, and then in the optimizer constructing phase, the code checks if this is NULL and if it is, it sets it to the metric corresponding to the neural networks loss? That sounds like the right plan.

EDIT:
**Oh no!** That really doesn't work, as the array for the metric function pointers are supposed to be stack allocated. If I then assign the array of functions pointer in a init function, that array will go out-of-scope when the function returns, and the world collapses, and the universe implodes and so on.

I need a better plan! 

... or maybe I can just make a `static` pointer to an array in a fixup function.... maybe that is a good idea?
